client/shared/src/main/scala/org/http4s/client/Client.scala

```diff
@@ -289,16 +291,57 @@ object Client {
       go(source).stream
     }
 
-    def runOrDispose(req: Request[F]): F[Resource[F, Response[F]]] =
-      Ref[F].of(false).map { disposed =>
-        val req0 = addHostHeaderIfUriIsAbsolute(req.pipeBodyThrough(until(disposed)))
-        Resource
-          .eval(app(req0))
-          .onFinalize(disposed.set(true))
-          .map(_.pipeBodyThrough(until(disposed)))
-      }
+    def run(
+        message: Message[F],
+        refOp: Option[Ref[F, Boolean]] = None,
+    ): Resource[F, Response[F]] = message match {
+      case req: Request[F] =>
+        Resource.eval(Ref[F].of(false)).flatMap { disposed =>
+          val reqAugmented =
+            addHostHeaderIfUriIsAbsolute(req.pipeBodyThrough(until(disposed)))
+          Resource
+            .eval(app(reqAugmented))
+            .onFinalize(disposed.set(true))
+            .flatMap(run(_, Option(disposed)))
+        }
+      case resp: Response[F] =>
+        for {
+          disposed <- Resource.eval(refOp.fold(Ref[F].of(false))(F.pure))
+          channel <- Resource.eval(Channel.synchronous[F, Chunk[Byte]])
+
+          _ <- resp.body.chunks
+            .through(channel.sendAll)
+            .compile
+            .drain
+            .uncancelable
+            .background
+
+          r = resp.withBodyStream(
+            Stream
+              .eval(disposed.get)
+              .ifM(
+                Stream.eval(F.raiseError(new IOException("response was disposed"))),
+                channel.stream.unchunks
+                  .onFinalize(
+                    channel.stream.compile.drain
+                      .guarantee(disposed.set(true))
+                  ),
+              )
+          )
+
+          _ <- Resource.onFinalize {
+            disposed.get
+              .ifM(
+                F.unit,
+                r.body.compile.drain,
+              )
+          }
+
+        } yield r
+
+    }
 
-    Client((req: Request[F]) => Resource.suspend(runOrDispose(req)))
+    Client((req: Request[F]) => run(req))
```

Comment:
Hmm, if we only pass `Request`s to `run`, why do we handle the `Message` then?

Reply:
Yes great eye! I noticed this as well :) I think it would be a lot clearer if we split it into two methods, instead of using a pattern match an `Option`. I didn't want to keep piling onto this PR 😅 but I would also really like to see this change!!